apps/epic-web/src/lib/transactions.ts

```diff
@@ -343,3 +342,279 @@ export async function fetchRefunds({
+    product: z.string().nullable(),
+    siteName: z.string().nullable(),
+    chargeId: z.string().nullable(),
+    amountRefunded: z.number().nullable(),
+  })
+
+export type CombinedBalanceTransaction = z.infer<
+  typeof CombinedBalanceTransactionSchema
+>
+
+export interface PaginatedCombinedBalanceTransactions {
+  transactions: CombinedBalanceTransaction[]
+  has_more: boolean
+  next_page_cursor: string | null
+}
+
+export async function fetchCombinedBalanceTransactions({
+  range,
+  start,
+  end,
+  limit = 100,
+  starting_after,
+}: z.infer<
+  typeof FetchBalanceTransactionsSchema
+>): Promise<PaginatedCombinedBalanceTransactions> {
+  try {
+    // Fetch balance transactions with pagination
+    const balanceTransactionsResult = await fetchBalanceTransactions({
+      range,
+      start,
+      end,
+      limit,
+      starting_after,
+    })
+
+    // Get unique charge and refund IDs from the current page of balance transactions
+    const chargeIds = new Set<string>()
+    const refundIds = new Set<string>()
+
+    balanceTransactionsResult.transactions.forEach((transaction) => {
+      if (transaction.type === 'charge' || transaction.type === 'payment') {
+        if (transaction.source) chargeIds.add(transaction.source)
+      } else if (
+        transaction.type === 'refund' ||
+        transaction.type === 'payment_refund'
+      ) {
+        if (transaction.source) refundIds.add(transaction.source)
+      }
+    })
+
+    // Fetch charges and refunds for the current page
+    const [charges, refunds] = await Promise.all([
+      Promise.all(
+        Array.from(chargeIds).map((id) => stripe.charges.retrieve(id)),
+      ),
+      Promise.all(
+        Array.from(refundIds).map((id) => stripe.refunds.retrieve(id)),
+      ),
+    ])
+
+    // Create maps for quick lookup
+    const chargeMap = new Map(
+      charges.map((charge) => [charge.id, simplifyCharge(charge)]),
+    )
+    const refundMap = new Map(
+      refunds.map((refund) => [refund.id, simplifyRefund(refund)]),
+    )
+
+    // Fetch associated charges for refunds
+    const refundCharges = await Promise.all(
+      refunds.map((refund) =>
+        refund.charge && typeof refund.charge === 'string'
+          ? stripe.charges.retrieve(refund.charge)
+          : null,
+      ),
+    )
+
+    // Update refundMap with charge information
+    refunds.forEach((refund, index) => {
+      const charge = refundCharges[index]
+      if (charge) {
+        const updatedRefund = simplifyRefund({
+          ...refund,
+          charge: charge as Stripe.Charge,
+        })
+        refundMap.set(refund.id, updatedRefund)
+      }
+    })
+
+    // Combined data balance transactions with charge and refund data
+    const combinedTransactions = balanceTransactionsResult.transactions.map(
+      (transaction): CombinedBalanceTransaction => {
+        if (
+          (transaction.type === 'charge' || transaction.type === 'payment') &&
```

Comment:
stripe added some methods of payments besides credit card, such as 'klarna' and it treats them as `payment` and not a charge - same with the refunds for those, so we need to add that type here.